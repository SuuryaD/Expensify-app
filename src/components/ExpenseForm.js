import React from 'react';
import 'react-dates/initialize';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';



export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            description: props.previous?props.previous.description: '',
            note: props.previous?props.previous.note:'',
            amount: props.previous?(props.previous.amount/100).toString():'',
            createdAt: props.previous?moment(props.previous.createdAt):moment(),
            calendarFocused: false,
            error: ''
        }

    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({
            description
        }))
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }) );
    }
    onAmountChange = (e) => {
        const amount = e.target.value;

        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(() => ({ amount }));
        }
    }
    onDateChange = (createdAt) => {
        this.setState(() => ({ createdAt }))
    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({calendarFocused: focused}))
    };
    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState(() => ({error: 'Please provide the details.'}))
        } else {
            this.setState(() => ({error: ''}))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    };


    render() {
        return (
            <div>
                
                {this.state.error !== '' && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        value={this.state.description}
                        onChange={this.onDescriptionChange} 
                    />
                    <input
                        type="number"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange} 
                    />
                    <SingleDatePicker 
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths = {1}
                        isOutsideRange={() => false}
                        id="aaaaaa"

                    />
                    <textarea
                        placeholder="Add a note to your expense"
                        value={this.state.note}
                        onChange = {this.onNoteChange}
                        
                    ></textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        )

    }
};