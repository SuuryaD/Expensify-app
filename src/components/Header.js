import React from 'react';
import {BrowserRouter , Route, Switch ,Link, NavLink} from 'react-router-dom';

const Header = () => (
    <div>
        <h1>Expensify App</h1>
        <NavLink to="/" activeClassName="is-active" exact={true}>Home </NavLink>
        <NavLink to="/create" activeClassName="is-active">create Expense </NavLink>
    </div>
);

export default Header;
