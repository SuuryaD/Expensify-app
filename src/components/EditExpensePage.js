import React from 'react';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(expense, this.props.expense.id);
        this.props.history.push('/');
    };
    onClick = () => {
        this.props.removeExpense(this.props.expense.id)
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <ExpenseForm 
                    previous={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button 
                    onClick={this.onClick}
                >Remove</button>
            </div>
        );
    }
};


const mapStateToProps = (state , props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
}
const mapDispatchToProps  = (dispatch) => {
    return {
        editExpense: (expense, id) => dispatch(editExpense(id, expense)),
        removeExpense: (id) => dispatch(removeExpense({ id }))
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);