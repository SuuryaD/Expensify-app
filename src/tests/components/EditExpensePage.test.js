import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let editExpense,removeExpense,history,wrapper;
beforeEach(() => {
    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = { push: jest.fn()};
    wrapper = shallow(<EditExpensePage editExpense={editExpense} history={history} removeExpense={removeExpense} expense={expenses[2]} />)
});

test('Should render edit expense page correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle edit expense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
    expect(editExpense).toHaveBeenLastCalledWith(expenses[2], expenses[2].id );
    expect(history.push).toHaveBeenLastCalledWith('/');
});

test('should handle remove expense', () => {
    wrapper.find('button').simulate('click');
    expect(removeExpense).toHaveBeenLastCalledWith(expenses[2].id);
    expect(history.push).toHaveBeenLastCalledWith('/');
});