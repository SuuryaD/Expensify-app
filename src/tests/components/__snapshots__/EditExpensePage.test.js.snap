// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render edit expense page correctly 1`] = `
<div>
  <ExpenseForm
    onSubmit={[Function]}
    previous={
      Object {
        "amount": 2000,
        "createdAt": 345600000,
        "description": "Credit Card",
        "id": "3",
        "note": "",
      }
    }
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
